### PR TITLE
Service mirror controllers should not create namespaces

### LINKED ISSUE #7565: Namespace creation on service mirroring conflicts with Infra-as-Code tooling
### What problem are you trying to solve?

When attempting to deploy using, e.g., Terraform or Pulumi, a Namespace with given labels and annotations on multiple clusters, one might declare a resource like so:

[Example code in practice](https://github.com/AaronFriel/frielforreal/blob/791befef776b824c9e6c986f58f25b7dde4567eb/node/infra/src/stacks/app-cockroachdb/stack.ts#L36-L44)

```ts
  const namespace = new k8s.core.v1.Namespace('example', {
    metadata: {
      name: 'example',
      annotations: { ... },
      labels: {
        'istio-injection': 'enabled',
        ...
      },
    },
  });
```

Unfortunately, IaC tools typically like to _own_ a resource's lifecycle. As a result, deploying the same namespace name to multiple clusters failed with an error in the IaC tool.

### How should the problem be solved?

An option to the service mirror to delay (block & retry) creation of resources when they are not cluster specific (having a namespace and/or resource name that is unique to a cluster).

In my case, I would like to deploy an app, such as `cockroachdb`, to the same namespace name in multiple clusters. However this results in conflicting ownership of the namespace and a race between the service mirror.

### Any alternatives you've considered?

We could try to upstream fixes to each IaC provider's code, but it's a thorny problem when a resource has multiple owners, e.g.: if each cluster attempted to provide a distinct annotation or label. ¯\\_(ツ)\_/¯ 

### How would users interact with this feature?

An option to the helm chart and/or CLI tool, much like `-enable-headless-services` is today.

### Would you like to work on this feature?

maybe

### PR DESCRIPTION
**Pull request by adleong**
Fixes #7565

When the service mirror controller detects an exported service in the target cluster, it attempts to create a mirror service for it in the source cluster in a namespace with the same name.  If that namespace does not exist in the source cluster, the service mirror controller will create it first.  This is somewhat unexpected behavior since the service mirror controller should not be in charge of administering cluster level resources such as namespaces.

We update the service mirror controller to instead log a warning and skip mirroring in this case.  Users who want to export a service should first ensure that the namespace exists in the source cluster.

Signed-off-by: Alex Leong <alex@buoyant.io>

### CODE REVIEW

**Comment by olix0r** [comment]
Will previously-observed services be created once the namespace is created? Can we test this?

**Comment by adleong** [comment]
Previously exported services won't be created immediately when the namespace is created, but will be the next time that service is updated.  So I'd recommend re-exporting a service after creating the namespace.

**Comment by olix0r** [comment]
@adleong how hard would it be to fix this so we update service s automatically? In a world where all of this information is stored in version control, toggling labels on services just to trigger re-indexing seems pretty cumbersome. I guess it would be easier to restart the mirror controller in this case; but I think it would be ideal if we could handle that state gracefully. It's not a hard blocker, but I think it's worth a follow-up to fix that.

File: multicluster/service-mirror/cluster_watcher.go [hunk 1]
```diff
@@ -490,8 +461,14 @@ func (rcsw *RemoteClusterServiceWatcher) handleRemoteServiceCreated(ctx context.
 	serviceInfo := fmt.Sprintf("%s/%s", remoteService.Namespace, remoteService.Name)
 	localServiceName := rcsw.mirroredResourceName(remoteService.Name)
 
-	if err := rcsw.mirrorNamespaceIfNecessary(ctx, remoteService.Namespace); err != nil {
-		return err
+	// Ensure the namespace exists, and skip mirroring if it doesn't
+	if _, err := rcsw.localAPIClient.NS().Lister().Get(remoteService.Namespace); err != nil {
+		if kerrors.IsNotFound(err) {
+			rcsw.log.Warnf("Skipping mirroring of service %s: namespace %s does not exist", serviceInfo, remoteService.Namespace)
```
Comment (mateiidavid) [inline multicluster/service-mirror/cluster_watcher.go:467]: Should we also emit a [`ServiceMirroringSkipped`](https://github.com/linkerd/linkerd2/blob/main/multicluster/service-mirror/cluster_watcher.go#L28) event, or would that complicate the code without adding too much value?

**Review by mateiidavid — APPROVED** [review]
Tested and works great!

**Review by olix0r — COMMENTED** [review]
How hard would it be to test out-of-order namespace creation?

File: multicluster/service-mirror/cluster_watcher.go [hunk 2]
```diff
@@ -517,6 +504,21 @@ func (rcsw *RemoteClusterServiceWatcher) handleRemoteServiceCreated(ctx context.
 	return rcsw.createGatewayEndpoints(ctx, remoteService)
 }
 
+func (rcsw *RemoteClusterServiceWatcher) handleLocalNamespaceAdded(ns *corev1.Namespace) error {
+	// When a local namespace is added, we issue a create event for all the services in the corresponding namespace in
+	// case any of them are exported and need to be mirrored.
```
Comment (kleimkuhler) [inline multicluster/service-mirror/cluster_watcher.go:509]: 👍

File: multicluster/service-mirror/cluster_watcher.go (hunk 1 above)
Comment (kleimkuhler) [reply multicluster/service-mirror/cluster_watcher.go:477]: I agree I think we should add a `rcsw.recorder.Event` here similar to when we [skip mirroring](https://github.com/linkerd/linkerd2/blob/9ad003c523394fd47fc8b2b0890cc0f1185b2e07/multicluster/service-mirror/cluster_watcher_headless.go#L50) headless services.

**Review by kleimkuhler — APPROVED** [review]
Overall looks good and works well!

I do think we should record an event—in addition to the log lines—for skipping mirroring.

**Review by olix0r — APPROVED** [review]
lgtm. I agree it would be good to emit events, but that seems like a fine follow-on change (with its own test? :)

**Pull request merged**